### ISSUE TITLE
Casting from long to int may cause wrong result

### ISSUE THREAD

**Issue by fduxiao** [pytorch/pytorch#50423]
## 🐛 Bug

Think about the following several lines. 
```C++
#include <iostream>
#include <torch/script.h>

using namespace std;

int main() 
{
    auto t = at::ones({1});
    auto p = t.to(at::kLong).to(at::kInt).data_ptr<int>();
    cout<<*p<<endl;
    return 0;
}

```

I expected this program to print `1`, but it sometimes gave results like `2055063114` or `0`. Each time I ran the compiled executable, I might received a different result. I think this only happens when I cast a tensor from long to int. 

Instead, if I use the following, it works well. 
```C++
#include <iostream>
#include <torch/script.h>

using namespace std;

int main() 
{
    auto t1 = at::ones({1});
    auto t2 = t1.to(at::kLong);
    auto t3 = t2.to(at::kInt);
    cout<<t1<<t2<<t3<<endl;
    auto p = t3.data_ptr<int>();
    cout<<*p<<endl;
    return 0;
}

```

I think this could be some memory management  error.

## Environment

libtorch is installed via Homebrew. The permalink is [here](https://github.com/Homebrew/homebrew-core/blob/68ec4844c08100660f10a864db751e1745c865f6/Formula/libtorch.rb).

PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.1 (x86_64)
GCC version: Could not collect
Clang version: 11.0.0
CMake version: version 3.19.2

Python version: 3.9 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.7.1
[pip3] torchaudio==0.7.2
[pip3] torchvision==0.8.2
[conda] Could not collect

cc @yf225 @glaringlee

**Comment by glaringlee** [comment]
@fduxiao Thanks for pointing this out, I will take a look at.

**Comment by glaringlee** [comment]
@fduxiao Hey, I think this is because your tensor got destroyed.
auto p = t.to(at::kLong).to(at::kInt).data_ptr<int>();
to() will create a new tensor, but the new tensor will be destroyed immediately since there is no reference to it.  Which means you call .data_ptr() and assign this ptr to p, so the newly created tensor has 0 ref count and destroyed immediately.

**Comment by fduxiao** [comment]
@glaringlee Thanks. I think I should write `auto p = *(t.to(at::kLong).to(at::kInt).data_ptr());`. Sorry for such a silly issue.

**Comment by nikitaved** [comment]
@fduxiao , do not forget to store Int Tensor in some intermediate variable like
```
auto int_tensor = ...to(at::kInt);
auto p = int_tensor.data_ptr<int>();
```

Otherwise p will point to a memory deallocated by destructor and you will get Segfault.

**Comment by nikitaved** [comment]
The code still will have an issue. You will need to use an intermediate Tensor variable like
```
auto int_t = t.to(at:kInt);
auto p = int_t.data_ptr<int>();
```

**Comment by fduxiao** [comment]
@nikitaved I see your point. If I don't store the intermediate tensor, the zero reference will cause it destroyed after I take the data_ptr. But shouldn't the destructor be called [after the full expression](https://en.cppreference.com/w/cpp/language/destructor), i.e., after the `;`? It seems a bit unnatural for me to always store an unnecessary variable and I think that's why the destructor is called later. 

I also disassembled the compiled binary. The pointer is dereferenced before the destructor.  
```asm
lea     rdi, [rbp+intermediate_int_tensor]
call    __ZNK2at6Tensor8data_ptrIiEEPT_v ; at::Tensor::data_ptr<int>(void)
mov     [rbp+data_ptr], rax
mov     rax, [rbp+data_ptr]
mov     ecx, [rax]
lea     rdi, [rbp+ intermediate_int_tensor] ; this
mov     [rbp+dereferenced_value_of_data_ptr], ecx
call    __ZN2at6TensorD1Ev ; at::Tensor::~Tensor()
```

In the release version, the dereference is done immediately after the call. 
```asm
lea     rdi, [rbp+ intermediate_int_tensor]
call    __ZNK2at6Tensor8data_ptrIiEEPT_v ; at::Tensor::data_ptr<int>(void)
mov     ebx, [rax]
```

Is it correct and idiomatic to write this in one line without the intermediate variable? I'm not familiar with modern C++ programming.

**Comment by nikitaved** [comment]
Have a look at the example from your reference: https://en.cppreference.com/w/cpp/language/destructor.
`to(at::kInt)` creates an intermediate Tensor which is not bound to any reference, because calling `data_ptr` on it does not affect its reference count. So it gets destroyed sooner or later, which results in unpredictable behavior: sometimes it segfaults, sometimes it works, sometimes you get some other memory corruption issues.
If you bind it to a reference, or call `operator=`, you bump its reference count so that the storage survives until the end of the scope and/or some program termination as explained in https://en.cppreference.com/w/cpp/language/destructor.

**Comment by fduxiao** [comment]
@nikitaved But the intermediate tensor should live throughout the whole expression. According to the [reference](https://en.cppreference.com/w/cpp/language/lifetime), `all temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created`, and a full-expression is an expression that is not a subexpression of another expression. That means when I dereference it, the tensor hasn't been destroyed.

**Comment by nikitaved** [comment]
It will live, yes, but the pointer will become dangling, no? You need `p` to point to some persistent memory.

**Comment by fduxiao** [comment]
@nikitaved Yes, the pointer is dangling, but I only need the integer value, which will be copied just well. In fact this tensor is used to obtain the predicted class from my network. I think C++ provides this lifetime to free programmers from always storing the temporary variables. 

Maybe a better idea is to write another member function returning the value instead of a pointer? I think it's safe to return such an integer (I think I won't do that for a higher dimensional tensor).

**Comment by nikitaved** [comment]
OK, what is your use case? I guess this is where the misunderstanding arises from.

**Comment by fduxiao** [comment]
@nikitaved My network gives a softmax output and I use `at::max` to obtain the index, which returns this tensor as a LongTensor. Then I cast it to int and use `data_ptr` to get the result. Maybe there's a better way to do this?

**Comment by nikitaved** [comment]
Hm, I am not sure why would you want to cast to `Int`. You can read a value of a singleton Tensor by calling `item<scalar_t>()` on it.
Also, think whether you really need that explicit index, there are plenty of methods from the advanced indexing sections like `index_*` which provide this functionality while being device-agnostic.

**Comment by fduxiao** [comment]
Em, I'm not sure either. I followed [this tutorial](https://pytorch.org/tutorials/advanced/cpp_export.html) translating the code from python to C++. I use `data_ptr` because I learned this from [converting a tensor from cv::Mat](https://discuss.pytorch.org/t/libtorch-c-convert-a-tensor-to-cv-mat-single-channel/47701). I have to export the result to an `enum class` so that others who do not know pytorch can make use of it. And the C++ documents never explain the functions provided in libtorch. Is there any good tutorial about how to use pytorch in production?

**Comment by nikitaved** [comment]
Not sure about production, but when it comes to advanced indexing in the cpp API, take a look at the following doc:
https://pytorch.org/cppdocs/notes/tensor_indexing.html

**Comment by fduxiao** [comment]
I see. I will follow your advice using `item`. Thanks a lot for helping me.

**Comment by glaringlee** [comment]
@fduxiao we were on holiday, late reply on this usage:
`auto p = *(t.to(at::kLong).to(at::kInt).data_ptr());`
This will work if you only need the int data once, and you are right no intermediate tensor will be destroyed during the evaluation of this expression. And all intermediate tensors will be destroyed after the expression evaluation finished, in opposite order of how they created. But, I don't recommend this usage since it will be hard to do full debug, and the behavior is fully depend on compiler (though C++ standard said this is ok, but not on your control). 
And if indexing can solve your problem, @nikitaved is right, see the tensor_indexing doc listed above.
